Expose sync height to gRPC interface

jkdufair opened the pull request:
This commit adds `sync_height` to the gRPC interface in order that
clients calculate progress while `lnd` syncs to the blockchain.

`sync_height` is exposed via the `GetInfo()` rpc call. Additionally,
`IsSynced()` returns the syncHeight as the second value in the tuple
that is returned, providing additional detail when querying about the
status of the sync. The `BtcWallet` interface has also been updated
accordingly.

This commit was created to support the issue to [Add progress bar for chain sync] (https://github.com/lightninglabs/lightning-app/issues/10) in lightning-app

jkdufair commented:
@Roasbeef Ok, that makes sense. I'll work on that (and update lightning-app accordingly. Should I expose it in the same location (return it from `IsSynced()`and expose it via `GetInfo()`. Is there any advantage to keeping `synced_height' in there as well?

Also, what do you mean by "w/e chain backed"?

jkdufair commented:
@Roasbeef Hopefully this is more of what you were looking for. I can also squash the commits if you think it's in the ballpark. And the name `wallet_best_branch_timestamp` may be a bit lengthy. I can shorten it if you think so.

Roasbeef commented on lnrpc/rpc.proto:
```diff
@@ -805,6 +805,9 @@ message GetInfoResponse {
     repeated string chains = 11 [ json_name = "chains" ];
+
+    /// Timestamp of the block best known to the wallet
+    int64 wallet_best_block_timestamp = 12 [ json_name = "wallet_best_block_timestamp" ];
```
Naming suggestion: `best_header_timestamp`.

Roasbeef commented on lnwallet/btcwallet/btcwallet.go:
```diff
@@ -656,14 +656,17 @@ func (b *BtcWallet) SubscribeTransactions() (lnwallet.TransactionSubscription, e
+	// also grab the timestamp of the best block to return for determining
+	// sync progress
 	syncState := b.wallet.Manager.SyncedTo()
+	walletBestBlock, err := b.chain.GetBlockHeader(&syncState.Hash)
```
The error isn't being checked here. Also you don't need to declare the error at the top of the scope.

Roasbeef commented on lnwallet/interface.go:
```diff
@@ -198,7 +198,9 @@ type WalletController interface {
 	// IsSynced returns a boolean indicating if from the PoV of the wallet,
 	// it has fully synced to the current best block in the main chain.
-	IsSynced() (bool, error)
+	// It also returns an int64 indicating the timestamp of the best block
```
Should note that the timestamp is expressed in unix epoch time.

Roasbeef commented on lnwallet/btcwallet/btcwallet.go:
```diff
@@ -656,14 +656,17 @@ func (b *BtcWallet) SubscribeTransactions() (lnwallet.TransactionSubscription, e
+	// also grab the timestamp of the best block to return for determining
+	// sync progress
 	syncState := b.wallet.Manager.SyncedTo()
+	walletBestBlock, err := b.chain.GetBlockHeader(&syncState.Hash)
```
Check out the bottom of this function: as is, we'd fetch the header for a _second_ unnecessarily. Also we can't directly reference the `chain` here as we need to fetch the header from the backend.

jkdufair replied:
That was my thought - to use the header that was already being fetched at the end of the function. That value seems not to change as the sync progresses. Which implies to me it's the header at the tip of the, i.e. btcd chain  and not the header at the tip of the wallet's chain. I just double checked and that does indeed seem to be the case.

Am I overlooking something?

Roasbeef commented:
Ah yeh, you're totally correct about that!

With that cleared up, this batch is looking pretty good. The final task left is to [re-structure the commits to match the contribution guide lines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#ModelGitCommitMessages) (note the portion about sub-system delimited commits).

jkdufair commented:
Thanks @Roasbeef. I rebased onto the current master and reworded the commit msg to match the guidelines. Since the changes affected two subsystems but were quite inter-related, I squashed them all and prefixed with both subsystems (I saw that in at least on other commit on `master`). If you want something different, I could, i.e. create two commits, one for `lnwallet` and one for `lnrpc`.

Thanks for your help :-)

jkdufair commented:
I think this is good to go after squashing and reformatting the commit messages? Not sure if I should rebase again to handle the conflicts? Let me know if there is anything else to do on my part.

Roasbeef commented:
Hey @jkdufair, the last thing that needs to be done is rebasing one final time to remove those conflicts. Thanks!

jkdufair commented:
@Roasbeef Done!

jkdufair commented on lnrpc/rpc.proto:
```diff
@@ -803,6 +803,9 @@ message GetInfoResponse {
     repeated string uris = 12 [json_name = "uris"];
+
+		/// Timestamp of the block best known to the wallet
+    int64 best_header_timestamp = 12 [ json_name = "best_header_timestamp" ];
```
Ok. I'll get it updated. Looks like other PRs keep stepping on this change.

Roasbeef commented:
Thanks for starting on this feature! I've rebased it manually and committed it as `84551c616f4968a27e66806cab7d55006a48487c`.
